### PR TITLE
sink(ticdc): fix table version used in storage sink

### LINKED ISSUE #8666: CDC storage sink data inconsistency when cdc scales and owner switch during workload running
### What did you do?

1. Create storage sink changefeed with below configurations, and run storage consumer to consumer the data to downstream mysql
```
[sink]
protocol='csv'
[sink.csv]
include-commit-ts=true
```

2. Run sysbench workload prepare
```
sysbench --db-driver=mysql --mysql-host=`nslookup upstream-tidb.cdc-testbed-tps-1683010-1-178 | awk -F: '{print $2}' | awk 'NR==5' | sed s/[[:space:]]//g`  --mysql-port=4000 --mysql-user=root --mysql-db=workload --tables=32 --table-size=100000 --create_secondary=off --debug=true --threads=32 --mysql-ignore-errors=2013,1213,1105,1205,8022,8027,8028,9004,9007,1062 oltp_write_only prepare
```
4. Run sysbench workload and at the same time do cdc scales and owner switch ( 3 > 6 -> owner switch - > 1 -> 5 -> owner swtich -> 2)
5. create a table in TiDB, and wait the table sync to downstream mysql (by consumer)
6. do data consistency check

### What did you expect to see?

Data consistency check should pass

### What did you see instead?

Data consistency check failed.

e.g. 
image

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
[root@upstream-tidb-0 /]# /tidb-server -V
Release Version: v7.0.0
Edition: Community
Git Commit Hash: 624aa887d74a2e6a28ec610b94918bbe25e48517
Git Branch: heads/refs/tags/v7.0.0
UTC Build Time: 2023-03-23 14:01:44
GoVersion: go1.20.2
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: unistore
```

Upstream TiKV version (execute `tikv-server --version`):

```console
│[root@upstream-tikv-0 /]# /tikv-server -V
│TiKV
│Release Version:   7.0.0
│Edition:           Community
│Git Commit Hash:   412614bf024808d6a45fec67b443df4e365a94eb
│Git Commit Branch: heads/refs/tags/v7.0.0
│UTC Build Time:    2023-03-23 12:15:00
│Rust Version:      rustc 1.67.0-nightly (96ddd32c4 2022-11-14)
│Enable Features:   pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raf
│t-engine cloud-aws cloud-gcp cloud-azure
│Profile:           dist_release
```

TiCDC version (execute `cdc version`):

```console
[root@upstream-ticdc-0 /]# /cdc version
Release Version: v7.0.0
Git Commit Hash: fd2efea950937113af39856d00670577745da576
Git Branch: heads/refs/tags/v7.0.0
UTC Build Time: 2023-03-23 12:02:40
Go Version: go version go1.20.2 linux/amd64
Failpoint Build: false
```

### PR DESCRIPTION
**Pull request by CharlesCheung96**
### What problem does this PR solve?

Issue Number: close #8666

### What is changed and how it works?
1. Set `TableInfoVersion` in [SingleTableTxn](https://github.com/pingcap/tiflow/pull/8751/files#diff-eb40c4235fa377719c71b4d416eb38d4f4134b281758dd43687ca525d0ca3549R669), which is [max{TableSinkStartTs, TableInfo.Version}](https://github.com/pingcap/tiflow/pull/8751/files#diff-021e72c7db585ee751d1c0c1985e4e552434699a2e2dcfa5afff9f86513a3225R121-R125).
2. Use correct `TableinfoVersion` in [TableDefinition](https://github.com/pingcap/tiflow/pull/8751/files#diff-ae808f9ebdd6f78efc57b1ee58e765b909c12be355cb2defffe7b44728ca6e44R198).
3. Use [loacl storage in integration test to facilitate debug](https://github.com/pingcap/tiflow/pull/8751/files#diff-0dac0ea75b05049ecd2a171ef67193c96891fc2677187257e06314dea92343d7R30).

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

##### Do you need to update user documentation, design documentation or monitoring documentation?

Need to update user documentation.

### Release note 

```release-note
`Fix the issue of table version rollback in storage sink when cdc scaling`.
```

### CODE REVIEW

**Comment by CharlesCheung96** [comment]
test infra verified: 1684987

File: cdc/model/sink.go [hunk 1]
```diff
@@ -664,11 +664,13 @@ func (d *DDLEvent) FromRenameTablesJob(job *model.Job,
...
 //msgp:ignore SingleTableTxn
 type SingleTableTxn struct {
-	Table     *TableName
-	TableInfo *TableInfo
-	StartTs   uint64
-	CommitTs  uint64
-	Rows      []*RowChangedEvent
+	Table            *TableName
+	TableInfo        *TableInfo
+	TableInfoVersion uint64
```
Comment (hi-rustin) [inline cdc/model/sink.go:669]: Could you please add a comment to it to explain how we get and use it?

Comment (CharlesCheung96) [reply cdc/model/sink.go:673]: Done, PTAL~

**Review by hicqu — APPROVED** [review]

File: cdc/sink/dmlsink/cloudstorage/dml_worker.go [hunk 2]
```diff
@@ -68,7 +68,7 @@ func newTableEventsMap() *tableEventsMap {
 }
 
 type wrappedTable struct {
-	tableName model.TableName
+	cloudstorage.VersionedTable
```
Comment (hi-rustin) [inline cdc/sink/dmlsink/cloudstorage/dml_worker.go:71]: Maybe we should call it VersionedTableName.

File: cdc/sink/dmlsink/event_appender.go [hunk 3]
```diff
@@ -103,16 +100,29 @@ func (t *TxnEventAppender) Append(
...
+	txn := &model.SingleTableTxn{
+		StartTs:   row.StartTs,
+		CommitTs:  row.CommitTs,
+		Table:     row.Table,
+		TableInfo: row.TableInfo,
+	}
+	if row.TableInfo != nil {
+		txn.TableInfoVersion = row.TableInfo.Version
+	}
+	if txn.TableInfoVersion < t.TableSinkStartTs {
```
Comment (hi-rustin) [inline cdc/sink/dmlsink/event_appender.go:124]: Could you please add a comment to explain what is going on here? Thanks!

File: deployments/ticdc/docker/storage-consumer.Dockerfile [hunk 4]
```diff
@@ -1,11 +1,4 @@
-FROM golang:1.20-alpine as builder
-RUN apk add --no-cache make bash
-WORKDIR /go/src/github.com/pingcap/tiflow
-COPY . .
-
-RUN make storage_consumer
-
 FROM alpine:3.15
-COPY --from=builder  /go/src/github.com/pingcap/tiflow/bin/cdc_storage_consumer /cdc_storage_consumer
+COPY ./cdc_storage_consumer /cdc_storage_consumer
```
Comment (hi-rustin) [inline deployments/ticdc/docker/storage-consumer.Dockerfile:2]: We can not copy the macOS version to the Linux image.

File: tests/integration_tests/canal_json_storage_basic/run.sh [hunk 5]
```diff
@@ -8,39 +8,13 @@ WORK_DIR=$OUT_DIR/$TEST_NAME
 CDC_BINARY=cdc.test
 SINK_TYPE=$1
 
-# start the s3 server
```
Comment (hi-rustin) [inline tests/integration_tests/canal_json_storage_basic/run.sh:11]: Can I ask why do we need to remove it?

Comment (CharlesCheung96) [reply tests/integration_tests/canal_json_storage_basic/run.sh:11]: we use [loacl storage in integration test to facilitate debug](https://github.com/pingcap/tiflow/pull/8751/files#diff-0dac0ea75b05049ecd2a171ef67193c96891fc2677187257e06314dea92343d7R30).

Comment (hi-rustin) [reply tests/integration_tests/canal_json_storage_basic/run.sh:11]: Got it. Thanks!

**Review by hi-rustin — APPROVED** [review]
Thanks!

**Pull request merged**